File: SaopParseExample/SaopParseExample/MiracSoapLib/ServiceHelper.m
//
//  ServiceHandler.m
//  SoapParse
//
//  Created by mirac on 27/06/16.
//  Copyright © 2016 asis. All rights reserved.


#import "ModelMaster.h"
#import "ServiceParameter.h"

@implementation ServiceHelper {
    BOOL ExecuteIsCompleted;
    NSMutableData *WebData;
    NSMutableArray *AuthValueArray;
    
    NSString *mStartElement;
    NSString *mDictionaryKey;
    NSString *mDictionaryLastValue;
    
    NSMutableDictionary *mResultDictionary;
    NSMutableString *mStringBuilder;
    
    NSMutableArray *mSRLineCoordinateArray;
    NSMutableArray *mSRTimeArray;

    
    
    NSString *mResult;
    NSString *mDescription;
    
    NSTimer *mTimer;
    
    
}

@synthesize MethodName;
@synthesize ParameterArray;

@synthesize ServiceURL;
@synthesize Namespace;
@synthesize SoapAction;

@synthesize callback;

-(id)initWithName:(NSString *)MethdName
{
    self = [super init];
    
    @try {
        
        
        mSRLineCoordinateArray = nil;
        mSRLineCoordinateArray = [[NSMutableArray alloc] init];
        
        mSRTimeArray = nil;
        mSRTimeArray = [[NSMutableArray alloc] init];
        
        
        ParameterArray = nil;
        ParameterArray = [[NSMutableArray alloc] init];
        

        
        mResult = @"0";
        mDescription = @"";
        
        self.MethodName = MethdName;
        //    NSURL *sRequestURL = [NSURL URLWithString:@"http://blablabla.com/Service.svc?wsdl"];
        self.ServiceURL = @"http://blablabla.com/A/B/C/service.svc";
        self.Namespace = @"http://tempuri.org/";
        self.SoapAction = [NSString stringWithFormat:@"http://tempuri.org/IWCFAbys/%@", MethodName];
    }
    @catch (NSException *exception) {
        
    }
    
    if (self) {
        
        
    }
    return self;
}

-(void)AddParameter:(NSString *)ParameterName ParameterValue:(NSString *)ParameterValue {
    
    @try {
        
        if (ParameterArray != nil) {
            
            ServiceParameter *mServiceParameter = [ServiceParameter new];
            
            mServiceParameter.ParameterName = ParameterName;
            mServiceParameter.ParameterValue = ParameterValue;
            
            [ParameterArray addObject: mServiceParameter];
        }
    }
    @catch (NSException *exception) {
        
    }
}


- (IBAction)TimerTick:(id)sender {
    
    @try {
        
        if (ExecuteIsCompleted == NO) {
            
            // Servis çağrısı henüz cevaplanmamış. Zaman aşımı nedeniyle işlem sonlandırılıyor.
            
            ServiceResult *mServiceResult = [ServiceResult new];
            
            mServiceResult.MethodName = @"Timeout";
            
            mServiceResult.Result = @"-273";
            mServiceResult.Description = @"Timeout.";
            
            [self.callback OnTaskCompleted:mServiceResult];
            
            ExecuteIsCompleted = YES;
            
            [mTimer invalidate];
        }
        
    }
    @catch (NSException *exception) {
        
    }
    
}




-(void)Execute {
    
    @try {
        
        NSMutableString *mSoapMessage = [[NSMutableString alloc] init];
        
        [mSoapMessage appendString: @"<?xml version=\"1.0\" encoding=\"utf-8\"?>"];
        [mSoapMessage appendString: @"<soap:Envelope xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\">"];
        [mSoapMessage appendString: @"<soap:Body>"];
        [mSoapMessage appendString: [NSString stringWithFormat:@"<%@ xmlns=\"%@\">", self.MethodName, self.Namespace]];
        
        if (self.ParameterArray != nil) {
            
            if (self.ParameterArray.count > 0) {
                
                for (ServiceParameter *mServiceParameter in self.ParameterArray) {
                    
                    [mSoapMessage appendString: [NSString stringWithFormat:@"<%@>%@</%@>", mServiceParameter.ParameterName, mServiceParameter.ParameterValue, mServiceParameter.ParameterName]];
                }
            }
        }
        
        [mSoapMessage appendString: [NSString stringWithFormat: @"</%@>", self.MethodName]];
        [mSoapMessage appendString: @"</soap:Body>"];
        [mSoapMessage appendString: @"</soap:Envelope>"];
        
        NSString *mMessage = [NSString stringWithString:mSoapMessage];
        
        NSURL *sRequestURL = [NSURL URLWithString:[NSString stringWithFormat: @"%@?wsdl", self.ServiceURL]];
        NSMutableURLRequest *myRequest = [NSMutableURLRequest requestWithURL:sRequestURL];
        NSString *sMessageLength = [NSString stringWithFormat:@"%lu", (unsigned long)[mMessage length]];
        
        [myRequest addValue: @"text/xml; charset=utf-8" forHTTPHeaderField:@"Content-Type"];
        [myRequest addValue: [NSString stringWithFormat: @"%@", self.SoapAction] forHTTPHeaderField:@"SOAPAction"];
        [myRequest addValue: sMessageLength forHTTPHeaderField:@"Content-Length"];
        [myRequest setHTTPMethod:@"POST"];
        [myRequest setHTTPBody: [mMessage dataUsingEncoding:NSUTF8StringEncoding]];
        
        NSURLConnection *theConnection = [[NSURLConnection alloc] initWithRequest:myRequest delegate:self];
        
        if( theConnection ) {
            WebData = [NSMutableData data];
        }else {
            NSLog(@"Some error occurred in Connection");
            
        }
        
    }
    @catch (NSException *exception) {
        
    }
    @finally {
        
    }
    
}

-(void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response{
    [WebData setLength:0];
}

-(void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data{
    [WebData appendData:data];
}

-(void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error{
    
    NSLog(@"Connection error. Please try again");
}

-(void)connectionDidFinishLoading:(NSURLConnection *)connection {
    
    NSLog(@"Received Bytes from Server: %lu", (unsigned long)[WebData length]);
    
    NSString *mXMLResponse = [[NSString alloc] initWithBytes:[WebData bytes] length:[WebData length] encoding:NSUTF8StringEncoding];
    NSData *mData = [mXMLResponse dataUsingEncoding:NSUTF8StringEncoding];
    [self parseXML:mData];
}

-(NSMutableArray *) parseXML:(NSData *)xmlData
{
    AuthValueArray = [NSMutableArray array];
    NSXMLParser *parser = [[NSXMLParser alloc] initWithData:xmlData];
    parser.delegate = self;
    [parser parse];
    
    return AuthValueArray;
}

-(void)parser:(NSXMLParser *)parser
didStartElement:(NSString *)elementName
 namespaceURI:(NSString *)namespaceURI
qualifiedName:(NSString *)qName
   attributes:(NSDictionary *)attributeDict {
    
    mStartElement = elementName;
}

-(void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string
{
    if ([mStartElement isEqualToString: mDictionaryKey]) {
        
        NSString *mPreviousValue = mDictionaryLastValue;
        mStringBuilder = [[mPreviousValue stringByAppendingString:string] mutableCopy];
        [mResultDictionary setObject:mStringBuilder forKey:mStartElement];
        mDictionaryLastValue= string;
        
    } else {
        
        [mResultDictionary setObject:string forKey:mStartElement];
        mDictionaryLastValue = string;
        mDictionaryKey = mStartElement;
    }
}

-(void)parser:(NSXMLParser *)parser
didEndElement:(NSString *)elementName
 namespaceURI:(NSString *)namespaceURI
qualifiedName:(NSString *)qName
{
    @try {

        
        if([self.MethodName isEqualToString: @"getREsponseTreeName"]){
            /**************************************************************************************/
            //Bayi Koordinatlari Listesi Result
            
//            Soap Metohod Name
            if([elementName isEqualToString:@"TestNethodName"]){
                
//                 like a tree in Response
            
                    TestObjectClass *mTestClass = [[TestObjectClass alloc] init];
                    
                    mTestClass.testString   = [mResultDictionary objectForKey:@"a:testObj"];
                    mTestClass.testString1  = [mResultDictionary objectForKey:@"a:testObj1"];
                    mTestClass.testString2  = [mResultDictionary objectForKey:@"a:testObj2"];
                    mTestClass.testString3  = [mResultDictionary objectForKey:@"a:testObj3"];
                    
                    [mSRTimeArray addObject:mTestClass];
                    
            }
            if([elementName isEqualToString:@"s:Envelope"]){
                
                ServiceResult *mServiceResult = [ServiceResult new];
                
                mServiceResult.MethodName = @"TestMethodName";
                
                mServiceResult.Description = @"items added in array.";
                mServiceResult.Result = @"Success";
                
                if (mSRTimeArray != nil) {
                    
                    if (mSRTimeArray.count > 0) {
                        
                        mServiceResult.mSRTimeArray = mSRTimeArray;
                    }
                }
                
                [self.callback OnTaskCompleted:mServiceResult];
            }
        }
        /**************************************************************************************/
        else if([self.MethodName isEqualToString: @"getCoordinatesMethodSoap"]){
            /**************************************************************************************/
            //getCoordinatesMethodSoap  Result
            
            if([elementName isEqualToString:@"Table1"]){
                TestObjectClass *mAnotherObject = [[TestObjectClass alloc] init];
                
                mAnotherObject.testString = [mResultDictionary objectForKey:@"responseName"]; //like a adress
                mAnotherObject.testString = [mResultDictionary objectForKey:@"responseName1"]; //like age "a:age or response name as "age ""
                mAnotherObject.testString = [mResultDictionary objectForKey:@"responseName2"];
                mAnotherObject.testString = [mResultDictionary objectForKey:@"responseName3"];

                
                
                [mSRLineCoordinateArray addObject:mAnotherObject];
                
            }
            if([elementName isEqualToString:@"s:Envelope"]){
                
                ServiceResult *mServiceResult = [ServiceResult new];
                
                mServiceResult.MethodName = @"getCoordinatesMethodSoap";
                
                mServiceResult.Description = @"items added in array.";
                mServiceResult.Result = @"Success";
                
                if (mSRLineCoordinateArray != nil) {
                    
                    if (mSRLineCoordinateArray.count > 0) {
                        
                        mServiceResult.SRLineCoordinateArray = mSRLineCoordinateArray;
                    }
                }
                
                [self.callback OnTaskCompleted:mServiceResult];
            }
        
        }
    }
    @catch (NSException *exception) {
        
    }
    
    
}
- (NSString *)StringValue:(NSString *)value StringSize:(NSInteger)count; {
    
    NSString *mResult = @"";
    
    if (value != nil) {
        
        if ([value length] < count) {
            
            for (int i = (int)[value length]; i < count; i++) {
                
                value = [NSString stringWithFormat:@"%@%@", value, @" "];
            }
        }
    }
    
    mResult = value;
    
    return  mResult;
}

@end

File: SoapParseLibSwift/SoapParseLibSwift/SaopLibMirac/ServiceHelper.swift
 //
 //  ServiceHelper.swift
 //  SoapParseLibSwift
 //
 //  Created by Asis on 16/10/2017.
 //  Copyright © 2017 Mirac. All rights reserved.
 //
 
 import UIKit
 
 class ServiceHelper: NSObject,XMLParserDelegate {
    
    var methodName                      = ""
    var serviceURL                      = ""
    var namespace                       = ""
    var soapAction                      = ""
    var mResult                         = ""
    var mDescription                    = ""
    var lastDictionaryValue : String    = ""
    var appendNewString : String        = ""
    var dictionaryKey : String          = ""
    var starterElement : NSString       = ""
    weak var callback: ServiceResultListener!
    
    var ExecuteIsCompleted = false
    var WebData: NSMutableData = NSMutableData()
    var WebData2  = NSMutableData()
    //    var mutableData : NSMutableData  = NSMutableData()
    var mResultDictionary = NSMutableDictionary()
    var mSRSmartStationRemaingTimeArray = NSMutableArray()
    
    var parameterArray                = NSMutableArray()
    var mSRLineCoordinateArray        = NSMutableArray()
    var mSRTimeArray                  = NSMutableArray()
    var AuthValueArray                = NSMutableArray()

    
    var mTimer: Timer?
    
    
    init (MethdName: String) {
        super.init()
        self.mSRSmartStationRemaingTimeArray  = NSMutableArray()
        
        self.parameterArray              = NSMutableArray()
        self.mSRLineCoordinateArray      = NSMutableArray()
        self.mSRTimeArray                = NSMutableArray()

        mResult      = ""
        mDescription = ""
        
        self.methodName = MethdName
        self.serviceURL = "http://blablabla.com/A/B/service.svc"
//        self.serviceURL = "http://blablabla.com/A/B/service.wsdl"
        self.namespace  = "http://tempuri.org/"
        self.soapAction = String(format:"http://tempuri.org/IWCFAbys/%@",methodName)
        
        
    }
    
    func addParameter(_ ParameterName: String, parameterValue ParameterValue: String) {
        if parameterArray != nil {
            let mServiceParameter : ServiceParameter = ServiceParameter()
            mServiceParameter.ParameterName = ParameterName
            mServiceParameter.ParameterValue = ParameterValue
            
            parameterArray.add(mServiceParameter)
            
        }
    }
    
    func TimerTick(){
        
        if (ExecuteIsCompleted == false){
            
            let mServiceResult : ServiceResult = ServiceResult()
            
            mServiceResult.MethodName = "Timeout"
            mServiceResult.Result = "-273"
            mServiceResult.Description = "Timeout"
            self.callback.OnTaskCompleted(serviceResult: mServiceResult)
            ExecuteIsCompleted = true
            mTimer?.invalidate()
        }
        
    }
    
    func execute() {
        
        let soapMessage : NSMutableString = ""
        soapMessage.append("<?xml version=\"1.0\" encoding=\"utf-8\"?>")
        soapMessage.append("<soap:Envelope xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\">")
        soapMessage.append("<soap:Body>")
        soapMessage.append(String(format:"<%@ xmlns=\"%@\">",self.methodName,self.namespace))
        
        if self.parameterArray != nil  {
            
            if self.parameterArray.count > 0{
                for mServiceParam in (parameterArray as NSArray as! [ServiceParameter])  {
                    
                    soapMessage.append(String(format:"<%@>%@</%@>",mServiceParam.ParameterName, mServiceParam.ParameterValue,mServiceParam.ParameterName))
                    //                    let mm: String = soapMessage as String!
                    //                    print(mm)
                }
            }
        }
        soapMessage.append(String(format:"</%@>",methodName))
        soapMessage.append("</soap:Body>")
        soapMessage.append("</soap:Envelope>")
        
        let mMessage: String = soapMessage as String!
        print(mMessage)
        
        let sRequestURL = URL(string : String(format: "%@?wsdl",serviceURL))
        let myRequest = NSMutableURLRequest(url: sRequestURL!)
        //        let sMessageLenght = String(format:"%lu",mMessage.characters.count)
        let sMessageLenght = mMessage.characters.count
        
        myRequest.httpBody = mMessage.data(using: String.Encoding.utf8,allowLossyConversion: false) // or false
        myRequest.addValue("text/xml; charset=utf-8", forHTTPHeaderField: "Content-Type")
        myRequest.addValue("\(soapAction)", forHTTPHeaderField: "SOAPAction")
        myRequest.addValue(String(sMessageLenght), forHTTPHeaderField: "Content-Length")
        myRequest.httpMethod = "POST"
        
        
        
        let connection = NSURLConnection(request: myRequest as URLRequest, delegate: self, startImmediately: true)!
        connection.start()
        if connection as NSURLConnection? != nil {
            //                WebData = [NSMutableData data];
            
            let WebData : Void = NSMutableData.initialize()
            print(WebData)
            
            //            WebData = Data() as? WebData ?? WebData
            //            WebData = NSMutableData()
            //            WebData = Data() as! NSMutableData    // Burası düzenlenecek
            //            WebData : Void  = NSMutableData.initialize()
        }
        else {
            print("Connection Problem")
        }
    }
    
    
    
    
    //    func connection(_ connection: NSURLConnection!, didReceiveResponse response: URLResponse!) {
    //        WebData.length = 0
    //    }
    //    func connection(_ connection: NSURLConnection!, didReceive data: Data!) {
    //        WebData.append(data)
    //    }
    //
    //    func connection(_ connection: NSURLConnection, didFailWithError error: Error) {
    //        print("Connection error. Please try again")
    //    }
    func connection(_ connection: NSURLConnection!, didReceiveResponse response: URLResponse!) {
        WebData.length = 0;
    }
    func connection(_ connection: NSURLConnection, didReceive challenge: URLAuthenticationChallenge) {
        print(WebData.length)
    }
    
    
    func connection(_ connection: NSURLConnection!, didReceiveData data: Data!) {
        WebData.append(data)
    }
    
    func connectionDidFinishLoading(_ connection: NSURLConnection!) {
        
        print (WebData.length)
        let mXMLResponse = NSString.init(bytes: WebData.bytes, length: WebData.length, encoding:String.Encoding.utf8.rawValue)
        let mData = mXMLResponse?.data(using: String.Encoding.utf8.rawValue)
        self.parseXML(xmlData: mData!)
    }
    
    func parseXML(xmlData: Data) -> NSMutableArray{
        let parser = XMLParser.init(data: xmlData)
        parser.delegate = self
        parser.parse()
        return AuthValueArray
    }
    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String]) {
        starterElement = elementName as NSString
    }
    func parser(_ parser: XMLParser, foundCharacters string: String) {
        
        if(starterElement .isEqual(to: self.dictionaryKey)){
            let ondekiString = self.lastDictionaryValue
            appendNewString = (ondekiString + string) //mutablecopy
            mResultDictionary.setObject(appendNewString, forKey: self.starterElement)
            self.lastDictionaryValue = string;
            
        }
        else {
            mResultDictionary.setObject(string, forKey: self.starterElement)
            self.lastDictionaryValue = string
            self.dictionaryKey = (self.starterElement as NSString) as String;
            
        }
    }
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        starterElement = elementName as NSString
        if ((methodName as NSString) .isEqual(to:"TestSoapMethod"))
        {
            if (starterElement .isEqual(to:"a:LineCoordinate")) {
                mResult = "1"
                mDescription = "Success"
                let mLatitude  : String  = mResultDictionary.object(forKey: "a:Latitude") as! String
                let mLongitude : String  = mResultDictionary.object(forKey: "a:Longitude") as! String
                if (mLatitude != "" && mLongitude != "" ){
                    
                    let mSRLine : LineObject = LineObject()
                    mSRLine.Latitute = mLatitude
                    mSRLine.Longitude = mLongitude
                    mSRLineCoordinateArray.add(mSRLine)
                }
            }
            if (starterElement .isEqual(to:"s:Envelope")){
                let mServiceResult : ServiceResult = ServiceResult()
                mServiceResult.MethodName = "TestSoapMethod"
                mServiceResult.Result = mResult
                mServiceResult.Description = mDescription
                if(mSRLineCoordinateArray.count != 0){
                    if(mSRLineCoordinateArray.count > 0){
                        mServiceResult.SRLineCoordinateArray = mSRLineCoordinateArray
                    }
                }
                self.callback.OnTaskCompleted(serviceResult: mServiceResult)
                
            }
            /**************************************************************************************/
            
        }

        else if (methodName as NSString).isEqual(to: "SoapTestMethod2"){
            /**************************************************************************************/
            /* CreateSessionMobileV2 Result */
            if(starterElement.isEqual(to:"a:SRSession")){
                let mSRSession : SessionClass = SessionClass()
                mSRSession.IsAlive     = (mResultDictionary.object(forKey: "a:isAlive")      as! String ).trimmingCharacters(in: CharacterSet.whitespaces)
                mSRSession.SessionId   = (mResultDictionary.object(forKey: "a:session_id")   as! String ).trimmingCharacters(in: CharacterSet.whitespaces)
                mSRSession.VersionCode = (mResultDictionary.object(forKey: "a:version_code") as! String ).trimmingCharacters(in: CharacterSet.whitespaces)
                
                mSRTimeArray.add(mSRSession)
            }
            if (starterElement .isEqual(to:"s:Envelope")){
                let mServiceResult : ServiceResult = ServiceResult()
                mServiceResult.MethodName = "SoapTestMethod2"
                mServiceResult.Result = mResult
                mServiceResult.Description = mDescription
                if(mSRTimeArray.count != 0){
                    if(mSRTimeArray.count > 0){
                        mServiceResult.SRTimerArray = mSRTimeArray;
                    }
                }
                if ExecuteIsCompleted == false {
                    ExecuteIsCompleted = true
                    mTimer?.invalidate()
                    self.callback.OnTaskCompleted(serviceResult: mServiceResult)
                }
                
            }
            /**************************************************************************************/
        }
        
    }
    
 }
 
